[wcf] Add support for non-SOAP based payloads

repl-chris opened the pull request:
Fixes #1204

## Changes
- Added compatibility for non-SOAP requests.
- Changed propagation to use both SOAP message headers and HTTP request headers - it will write to both and will favor HTTP headers when reading, to support transport-level instrumentation.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

repl-chris commented:
This PR is blocked waiting for https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1247

mconnew commented on src/OpenTelemetry.Instrumentation.Wcf/TelemetryPropagationWriter.cs:
```diff
@@ -0,0 +1,105 @@
+            return;
+        }
+
+        request.Headers.Add(MessageHeader.CreateHeader(name, "https://www.w3.org/TR/trace-context/", value, false));
```
This overload instantiates a new DataContractSerializer instance every time the created header is written to the outgoing SOAP message. You can either cache a static instance of DataContractSerializer created for serializing String and use a different CreateHeader overload which takes XmlObjectSerializer as a parameter (DCS derives from XmlObjectSerializer), or create your own header type that derives from MessageHeader and implement the OnWriteHeaderContents method to call XmlDictionaryWriter.WriteString. The latter option will have slightly better perf, but not by a lot as a string DCS is relatively light weight.

mconnew commented on src/OpenTelemetry.Instrumentation.Wcf/TelemetryPropagationReader.cs:
```diff
@@ -0,0 +1,109 @@
+        var headerIndex = request.Headers.FindHeader(name, "https://www.w3.org/TR/trace-context/");
+        return headerIndex < 0
+            ? null
+            : new[] { request.Headers.GetHeader<string>(headerIndex) };
```
This also instantiates a DataContractSerializer every time it's used. Either cache the DCS and use a different overload of GetHeader where you can provide the serializer, or call Headers.GetReaderAtHeader to get the XmlDictionaryReader for the header, then call ReadContentAsString to read the value. You might have to call `Read` to move forward in the state machine before you can call ReadContentAsString, but I'm not sure. You should wrap in a try/catch(XmlException) to be safe in case of a malformed header.

repl-chris commented:
@mconnew Thanks for the review 😀

I have addressed your comments by manually implementing the xml reading/writing instead of relying on the datacontractserializers.  I have also extracted the needed test pieces from 1247 so this one is no longer blocked, since it looks like that one will probably take a little longer.

mconnew commented on src/OpenTelemetry.Instrumentation.Wcf/TelemetryPropagationReader.cs:
```diff
@@ -0,0 +1,108 @@
+    /// <param name="request">The incoming <see cref="Message"/> to read trace context from.</param>
+    /// <param name="name">The header name being requested.</param>
+    /// <returns>An enumerable of all the values for the requested header, or null if the requested header was not present on the request.</returns>
+    public static IEnumerable<string> SoapMessageHeaders(Message request, string name)
```
Conceptually why is this returning an IEnumerable when both the HTTP and SOAP Header mechanism only allow for a single instance item?

repl-chris replied:
It only returns IEnumerable because the `TextMapPropagator.Extract()` method expects IEnumerable (see https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Context/Propagation/TextMapPropagator.cs#L50). I don't really understand that use-case but I figured it must be like that on the propagators for a reason, so I would just maintain that capability.

repl-chris replied:
Ok, today I learned - the reason is baggage. The value for baggage is unbounded and can overflow a single entry, so this interface gives opentelemetry the ability to have full control over how to split the value across multiple entries and subsequently reconstitute it on the receiving side.  It will call the writer multiple times with the same header name - once for each part of the split-up value....and then on the receiving side it will just call the reader once to retrieve all the parts from the carrier at once.  This way the readers+writers don't need to implement any of the splitting logic, they just need to support reading and writing multiple values for the same header name.  So, I guess, the interface is correct but the implementation will not support large baggage values. This new implementation, however, has the same baggage support as the existing implementation in `main` so I'd rather not tackle "proper baggage support" as part of this PR, I'd rather do it in a separate PR.

Kielek commented:
@repl-chris, your previous PR is finally merged. Could you please resolve conflicts? I hope we will be able to proceed with this a bit faster :)

repl-chris commented:
@Kielek Conflicts resolved, good to go 😀

Kielek commented:
@CodeBlanch. @mconnew, could you please make the final review?

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/WcfInstrumentationOptions.cs:
```diff
@@ -62,4 +63,28 @@ public class WcfInstrumentationOptions
+    /// <para>string: The header name being requested.</para>
+    /// <para>IEnumerable&lt;string&gt; (return): All the values for the requested header, or null if the requested header was not present on the request.</para>
+    /// </remarks>
+    public Func<Message, string, IEnumerable<string>> PropagationReader { get; set; } = TelemetryPropagationReader.Default;
```
@repl-chris What is the use case for users needing to mess with these new options? If it isn't super strong and the defaults will work for most users, let's make them `internal`. We can always expose them later if there is a strong need.

repl-chris replied:
The main reason I exposed these is because the feature was requested here https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/209 ....but I wouldn't consider that a "strong need".  Theoretically users could be using a transport that the defaults will not work for, but I would expect the defaults to work for the vast majority of users.

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@
   ([#1255](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1255))
 * Client instrumentation implementation moved to lower-level `BindingElement`.
   ([#1247](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1247))
+* Added support for non-SOAP requests.
+  ([#1251](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1251))
```
@repl-chris Sorry I probably created a merge conflict for you by releasing the other PR on #1308 would you mind updating this file and moving the new entry up into the "Unreleased" section?

repl-chris replied:
updated 👍

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/OpenTelemetry.Instrumentation.Wcf.csproj:
```diff
@@ -20,6 +20,7 @@
 
   <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
     <PackageReference Include="System.ServiceModel.Primitives" Version="4.7.0" />
+    <PackageReference Include="System.ServiceModel.Http" Version="4.7.0" />
```
Mixed feelings about adding this reference. Not sure if anyone will complain. Seems like we only need `HttpRequestMessageProperty` what if we accessed it reflectively and only performed the http actions if we detect the class/assembly is there?

If we do keep the ref, it looks like `System.ServiceModel.Http` pulls in `System.ServiceModel.Primitives` so you could probably remove that explicit reference.

Semi-related, how do you feel about updating the README to explicitly say we have tested this package with SOAP + HTTP transports and they will work but others may not?

repl-chris replied:
I like that idea, no need to bring in the reference if the application isn't using HTTP.  Done 👍

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/Implementation/HttpRequestMessagePropertyWrapper.cs:
```diff
@@ -0,0 +1,102 @@
+            }
+        }
+        catch (Exception)
+        {
```
nit: Might be worth adding an EventSource log here to help troubleshoot any oddities

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/Implementation/HttpRequestMessagePropertyWrapper.cs:
```diff
@@ -0,0 +1,102 @@
+
+    public object InnerObject => this.innerObject;
+
+    private static ReflectedInfo Initialize()
```
@repl-chris This is nice thanks for doing it! A little bit of feedback...

* `Name` seems to be static so we could probably just cache the result?
* `Headers` we are invoking through the `MethodInfo` which is pretty slow. It is better to create a delegate from that. We have this class `PropertyFetcher` you can link into the project which will help with that. I have some benchmarks below that show the improvement. Doing this exercise I noticed the perf of `PropertyFetcher` is kind of crappy. On the benchmarks I have a lower-level version which is even better. I think though, use `PropertyFetcher` and we can improve its performance which will then benefit everything else?

|                              Method |      Mean |     Error |    StdDev | Allocated |
|------------------------------------ |----------:|----------:|----------:|----------:|
|      RetrieveHeadersUsingReflection | 56.310 ns | 0.8513 ns | 0.7546 ns |         - |
|        RetrieveHeadersUsingDelegate |  6.571 ns | 0.1724 ns | 0.1917 ns |         - |
| RetrieveHeadersUsingPropertyFetcher | 16.139 ns | 0.2702 ns | 0.2527 ns |         - |

```csharp
[MemoryDiagnoser]
public class WcfReflectionBenchmarks
{
    private static readonly ReflectedInfo ReflectedValues = Initialize();

    private readonly HttpRequestMessageProperty request = new();

    [Benchmark]
    public WebHeaderCollection RetrieveHeadersUsingReflection()
    {
        return (WebHeaderCollection)ReflectedValues.HeadersProperty.GetMethod.Invoke(this.request, null);
    }

    [Benchmark]
    public WebHeaderCollection RetrieveHeadersUsingDelegate()
    {
        return ReflectedValues.GetHeaders.Invoke(this.request);
    }

    [Benchmark]
    public WebHeaderCollection RetrieveHeadersUsingPropertyFetcher()
    {
        return ReflectedValues.GetHeadersPropertyFetcher.Fetch(this.request);
    }

    private static ReflectedInfo Initialize()
    {
        var type = Type.GetType(
            "System.ServiceModel.Channels.HttpRequestMessageProperty, System.ServiceModel, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089",
            true);
        var headersProp = type.GetProperty("Headers", BindingFlags.Public | BindingFlags.Instance, null, typeof(WebHeaderCollection), Array.Empty<Type>(), null);
        var nameProp = type.GetProperty("Name", BindingFlags.Public | BindingFlags.Static, null, typeof(string), Array.Empty<Type>(), null);

        if (type != null && headersProp != null && nameProp != null)
        {
            return new ReflectedInfo
            {
                Type = type,
                Name = (string)nameProp.GetMethod.Invoke(null, null),
                HeadersProperty = headersProp,
                GetHeaders = ReflectedInfo.CreateDelegateWrapper<WebHeaderCollection>(headersProp.GetMethod),
                GetHeadersPropertyFetcher = new PropertyFetcher<WebHeaderCollection>("Headers"),
            };
        }

        throw new NotSupportedException("WCF reflection failed to bind");
    }

    private sealed class ReflectedInfo
    {
        public Type Type;
        public string Name;
        public PropertyInfo HeadersProperty;
        public DelegateWrapper<WebHeaderCollection> GetHeaders;
        public PropertyFetcher<WebHeaderCollection> GetHeadersPropertyFetcher;

        public static DelegateWrapper<T> CreateDelegateWrapper<T>(MethodInfo getMethodInfo)
        {
            var delegateWrapperStronglyTyped = Activator.CreateInstance(
                typeof(DelegateWrapperStronglyTyped<,>).MakeGenericType(getMethodInfo.DeclaringType, getMethodInfo.ReturnType),
                BindingFlags.Public | BindingFlags.Instance,
                binder: null,
                args: new object[] { getMethodInfo },
                culture: null);

            return (DelegateWrapper<T>)delegateWrapperStronglyTyped;
        }

        public abstract class DelegateWrapper<T>
        {
            public abstract T Invoke(object instance);
        }

        private sealed class DelegateWrapperStronglyTyped<TClass, TProperty> : DelegateWrapper<TProperty>
        {
            private readonly Func<TClass, TProperty> innerDelegate;

            public DelegateWrapperStronglyTyped(MethodInfo getMethodInfo)
            {
                this.innerDelegate = (Func<TClass, TProperty>)getMethodInfo.CreateDelegate(typeof(Func<TClass, TProperty>));
            }

            public override TProperty Invoke(object instance)
                => this.innerDelegate((TClass)instance);
        }
    }
}
```

Kielek replied:
Worth to check if we can just grab newer version PropertyFetcher from main repository. I suppose it was not synced for a while.

repl-chris replied:
I have updated the reflection as suggested.  I looked into grabbing the newer version of PropertyFetcher, but it has breaking changes so I backed away slowly...probably should do that in a separate PR.  

On a side note, anyone have any tips/tricks on how to easily test those EventSource logs?...I was able to manually test the new one, but it took me way too long to figure out and I'm sure there must be an easier way than what I did 😀

Kielek replied:
I have created #1311 to track this update. It will be great if you would like to contribute. For sure, it is not a blocker for this PR.

CodeBlanch replied:
@repl-chris RE: The EventSource what you want to do? Set up a listener? This class might help you with that: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/test/OpenTelemetry.Contrib.Tests.Shared/TestEventListener.cs

IIRC @TimothyMothra has some experience testing EventSources. Depending on what you want to do he may be able to help as well.

repl-chris replied:
Oooooh yeah, that's exactly what I didn't know I was looking for...I've never played with ETW events before so I was running around in the dark bumping into things 😂

CodeBlanch replied:
@repl-chris Hey sorry I haven't approved/merged this because it sounded like you were going to do some kind of test for the EventSource. I don't think it is super important that we do that, but if you want to by all means feel free. Just LMK if you are all done.

repl-chris commented:
@CodeBlanch Is there anything else you'd like done for this PR?

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/Implementation/TelemetryPropagationReader.cs:
```diff
@@ -0,0 +1,107 @@
+            return null;
+        }
+
+        var value = new HttpRequestMessagePropertyWrapper(prop).Headers[name];
```
Why force an allocation here (`new HttpRequestMessagePropertyWrapper`)? Seems like this could all be done statically.

```csharp
static class HttpRequestMessagePropertyHelper // Renamed + made the whole class static
{
   public static WebHeaderCollection? GetHeaders(object? httpRequestMessageProperty)
   {
          EnsureEnabled();

          if (httpRequestMessageProperty == null)
             return null; // Could return new() here, or could return a static empty instance if there is a way to make it readonly (not sure about that). Null is cheap

         return ReflectedValues.HeadersFetcher.Fetch(httpRequestMessageProperty);
   } 

   public static bool TryGetHeaderByName(object? httpRequestMessageProperty, string name, out string? value)
   {
        EnsureEnabled();

        if (httpRequestMessageProperty != null)
        {
              var headers = ReflectedValues.HeadersFetcher.Fetch(httpRequestMessageProperty);
              value = headers[name];
              return value != null;
        }

        value = null;
        return false;
   }
}
```

repl-chris replied:
Sounds good...I've updated it to be all static

CodeBlanch commented on src/OpenTelemetry.Instrumentation.Wcf/Implementation/HttpRequestMessagePropertyWrapper.cs:
```diff
@@ -0,0 +1,111 @@
+        return null;
+    }
+
+    private static void EnsureEnabled()
+    {
+        if (!IsHttpFunctionalityEnabled)
+        {
+            throw new InvalidOperationException("Http functionality is not enabled, check IsHttpFunctionalityEnabled before calling this method");
+        }
+    }
```
It seems like the calling code on the PR is checking `IsHttpFunctionalityEnabled` correctly so is the goal here to catch potential mistakes at dev time? Could we move under a debug block maybe?

```csharp
    private static void EnsureEnabled()
    {
#if DEBUG
        if (!IsHttpFunctionalityEnabled)
        {
            throw new InvalidOperationException("Http functionality is not enabled, check IsHttpFunctionalityEnabled before calling this method");
        }
#endif
    }
```

repl-chris replied:
I've refactored this safety check and the object type check to be DEBUG-only 👍

CodeBlanch reviewed (approved):
LGTM
